replica/database.cc

```diff
@@ -3184,4 +3184,112 @@ database::on_effective_service_levels_cache_reloaded() {
     co_return;
 }
 
+utils::chunked_vector<uint64_t> compute_random_sorted_ints(uint64_t max_value, uint64_t n_values) {
+    static thread_local std::minstd_rand rng{std::random_device{}()};
+    std::uniform_int_distribution<uint64_t> dist(0, max_value);
+    utils::chunked_vector<uint64_t> chosen;
+    chosen.reserve(n_values);
+    for (size_t i = 0; i < n_values; ++i) {
+        chosen.push_back(dist(rng));
+    }
+    std::ranges::sort(chosen);
+    return chosen;
+}
+
+future<utils::chunked_vector<bytes>> database::sample_data_files(
+    table_id id,
+    uint64_t chunk_size,
+    uint64_t n_chunks
+) {
+    auto ticket = get_units(_sample_data_files_sharded_concurrency_limiter, 1);
+
+    struct state_by_shard {
+        utils::chunked_vector<sstables::sstable_files_snapshot> snapshot;
+        schema_ptr schema;
+        uint64_t local_chunks;
+        uint64_t cross_shard_offset;
+    };
+    std::vector<foreign_ptr<std::unique_ptr<state_by_shard>>> state(smp::count);
+
+    co_await container().invoke_on_all(coroutine::lambda([&] (auto& local_db) -> future<> {
+        auto& local_state = state[this_shard_id()];
+        local_state = make_foreign(std::make_unique<state_by_shard>());
+        auto t = local_db.get_tables_metadata().get_table_if_exists(id);
+        if (!t) {
+            throw std::runtime_error(fmt::format("sample_data_files: table {} does not exist", id));
+        }
+        local_state->schema = t->schema();
+        local_state->snapshot = co_await t->take_sstable_set_snapshot();
```

Comment:
I think you're fine with just using table::get_sstables(), take_sstable_set_snapshot() is more expensive, since it opens a file desc for every sstable component, and only needed when you need to find file descs for all components that might have been deleted already. You don't seem to have this requirement since you only want to read chunks, and all data is available through sstable object.

Reply:
True.

I will change `take_sstable_set_snapshot()` to copy only the `shared_sstable` objects without opening other components.

(I have to make a copy because AFAICS `table::get_sstables()` returns a shared pointer to a set that can change under my feet if I don't copy it).

Reply:
Done.